docs/topics/object-declarations.md

```diff
@@ -193,32 +193,92 @@ fun main() {
+{type="warning"}
 
 ```kotlin
-sealed class ReadResult {
-    data class Number(val value: Int): ReadResult()
-    data class Text(val value: String): ReadResult()
-    data object EndOfFile: ReadResult()
+import java.lang.reflect.Constructor
+
+data object MySingleton
+
+fun main() {
+    val evilTwin = createInstanceViaReflection()
+
+    println(MySingleton) // MySingleton
+    println(evilTwin) // MySingleton
+
+    // Even when a library forcefully creates a second instance of MySingleton, its `equals` method returns true:
+    println(MySingleton == evilTwin) // true
+
+    // Do not compare data objects via ===.
+    println(MySingleton === evilTwin) // false
+}
+
+fun createInstanceViaReflection(): MySingleton {
+    // Kotlin reflection does not permit the instantiation of data objects.
+    // This creates a new MySingleton instance "by force" (i.e. Java platform reflection)
+    // Don't do this yourself!
+    return (MySingleton.javaClass.declaredConstructors[0].apply { isAccessible = true } as Constructor<MySingleton>).newInstance()
+}
+```
+
+The generated `hashCode()` function has behavior that is consistent with the `equals()` function, so that all runtime instances of a `data object` have the same hash code.
+
+#### No copy and componentN functions
+
+While `data object` and `data class` declarations are often used together and have some similarities, there are some functions that are not generated for a `data object`:
+
+- No `copy()` function: Because a `data object` declaration is intended to be used as singleton objects, no `copy()` function is generated. The singleton pattern restricts the instantiation of a class to one, single instance, which would be violated by allowing copies of the instance to be created.
+- No `componentN()` function: unlike a `data class`, a `data object` does not have any data properties. Since attempting to destructure such an object without data properties would not make sense, no `componentN()` functions are generated.
+
+#### Use cases for data objects
+
+`data object` declarations are a particularly good fit for sealed hierarchies, like [sealed classes or sealed interfaces](sealed-classes.md), since they allow you to maintain symmetry with any data classes you might have defined alongside the object:
+
+```kotlin
+sealed interface ReadResult
+data class Number(val number: Int) : ReadResult
+data class Text(val text: String) : ReadResult
+data object EndOfFile : ReadResult
+
+fun printReadResult(r: ReadResult) {
+  when(r) {
+    is Number -> println("Num(${r.number}")
+    is Text -> println("Txt(${r.text}")
+    is EndOfFile -> println("EOF")
+  }
 }
 
 fun main() {
-    println(ReadResult.Number(1)) // Number(value=1)
-    println(ReadResult.Text("Foo")) // Text(value=Foo)
-    println(ReadResult.EndOfFile) // EndOfFile
+  printReadResult(EndOfFile) // EOF
 }
 ```
 
+
+
```

Comment:
```suggestion

```

Reply:
Seems two of these newline changes were conflicting after applying them in order. If you still need changes here, do take a look :)